libvast/src/system/importer.cpp

```diff
@@ -38,44 +39,72 @@ importer_state::importer_state(caf::event_based_actor* self_ptr)
 }
 
 importer_state::~importer_state() {
-  write_state();
+  write_state(write_mode::with_next);
 }
 
 caf::error importer_state::read_state() {
-  auto file = dir / "next_id";
+  auto file = dir / "current_id_block";
   if (exists(file)) {
     VAST_VERBOSE(self, "reads persistent state from", file);
-    std::ifstream available{to_string(file)};
-    available >> next_id;
-    if (!available.eof())
-      VAST_ERROR(self, "got an invalidly formatted persistence file:", file);
+    std::ifstream state_file{to_string(file)};
+    state_file >> current.end;
+    state_file >> current.next;
+    if (!state_file) {
+      VAST_WARNING(self, "did not find next id position from the state file; "
+                         "detected an irregular shutdown");
+      current.next = current.end;
+    }
   } else {
     VAST_VERBOSE(self, "did not find a state file at", file);
+    current.end = 0;
+    current.next = 0;
   }
-  return caf::none;
+  return get_next_block();
 }
 
-caf::error importer_state::write_state() {
+caf::error importer_state::write_state(write_mode mode) {
   if (!exists(dir)) {
     auto result = mkdir(dir);
     if (!result)
       return std::move(result.error());
   }
-  std::ofstream available{to_string(dir / "next_id")};
-  available << next_id;
-  VAST_VERBOSE(self, "persisted id space caret at", next_id);
+  std::ofstream state_file{to_string(dir / "current_id_block")};
+  state_file << current.end;
+  if (mode == write_mode::with_next) {
+    state_file << " " << current.next;
+    VAST_VERBOSE(self, "persisted id block [", current.next, ",", current.end,
+                 ")");
+  } else {
+    VAST_VERBOSE(self, "persisted id block boundary at", current.end);
+  }
   return caf::none;
 }
 
+caf::error importer_state::get_next_block() {
+  using namespace si_literals;
+  while (current.next >= current.end)
+    current.end += 8_Mi;
+  return write_state(write_mode::without_next);
+}
+
+id importer_state::next_id(uint64_t advance) {
+  id pre = current.next;
+  current.next += advance;
+  if (current.next >= current.end)
+    get_next_block();
+  return pre;
+}
+
 id importer_state::available_ids() const noexcept {
-  return max_id - next_id;
+  return max_id - current.next;
 }
 
 caf::dictionary<caf::config_value> importer_state::status() const {
   caf::dictionary<caf::config_value> result;
   // Misc parameters.
   result.emplace("available-ids", available_ids());
-  result.emplace("next-id", next_id);
+  result.emplace("next-id", current.next);
+  result.emplace("block-boundary", current.end);
```

Comment:
This could be wrapped into a dictionary with key `ids`; maybe this `.` notation even works:

```suggestion
  result["ids.available"] = available_ids();
  result["ids.block.next"] = current.next;
  result["ids.block.end"] = current.end;
```